File: client/client-multi/client-cpp/impl/common/EndpointObjectHash.cpp
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <vector>

#include <botan/sha160.h>

#include "kaa/common/EndpointObjectHash.hpp"
#include "kaa/common/exception/KaaException.hpp"

namespace kaa {

EndpointObjectHash::EndpointObjectHash(const std::uint8_t *data, const std::uint32_t &dataSize)
{
    calculateHash(data, dataSize);
}

EndpointObjectHash::EndpointObjectHash(const std::string& str)
{
    calculateHash(reinterpret_cast<const std::uint8_t *>(str.c_str()), str.size());
}

EndpointObjectHash::EndpointObjectHash(const SharedDataBuffer& endpointHash)
{
    calculateHash(endpointHash.first.get(), endpointHash.second);
}

EndpointObjectHash::EndpointObjectHash(const EndpointObjectHash& endpointHash) : hashDigest_(endpointHash.hashDigest_)
{

}

EndpointObjectHash::EndpointObjectHash(EndpointObjectHash&& endpointHash) : hashDigest_(std::move(endpointHash.hashDigest_))
{

}

EndpointObjectHash& EndpointObjectHash::operator=(const EndpointObjectHash& endpointHash)
{
    hashDigest_ = endpointHash.hashDigest_;
    return *this;
}

EndpointObjectHash& EndpointObjectHash::operator=(EndpointObjectHash&& endpointHash)
{
    hashDigest_ = std::move(endpointHash.hashDigest_);
    return *this;
}

std::vector<std::uint8_t> EndpointObjectHash::getHashDigest()
{
    return hashDigest_;
}

void EndpointObjectHash::calculateHash(const std::uint8_t* data, std::uint32_t dataSize)
{
    if (!data && dataSize != 0) {
        throw KaaException("empty raw data or null size");
    }
    Botan::SHA_160 sha;
    const auto& result = sha.process(data, dataSize);
    hashDigest_.assign(result.begin(), result.end());
}

EndpointObjectHash::operator std::vector<std::uint8_t>()
{
    return hashDigest_;
}

} /* namespace kaa */

File: common/endpoint-shared/src/main/java/org/kaaproject/kaa/common/hash/EndpointObjectHash.java
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kaaproject.kaa.common.hash;

import org.apache.commons.codec.binary.Base64;

import java.io.Serializable;
import java.nio.ByteBuffer;
import java.nio.charset.Charset;
import java.util.Arrays;


/**
 * The class EndpointObjectHash is responsible for hash calculation
 *
 * @author Andrew Shvayka
 */
public final class EndpointObjectHash implements Serializable {

    private static final Charset UTF8 = Charset.forName("UTF-8");

    private static final long serialVersionUID = 1L;

    private final byte[] data;

    /**
     * Instantiates a new endpoint object hash.
     *
     * @param data the data
     */
    private EndpointObjectHash(byte[] data) {
        super();
        this.data = Arrays.copyOf(data, data.length);
    }

    /**
     * From bytes.
     *
     * @param data the data
     * @return the endpoint object hash
     */
    public static EndpointObjectHash fromString(String data) {
        if (data == null) {
            return null;
        }
        return new EndpointObjectHash(Base64.encodeBase64(data.getBytes(UTF8)));
    }

    /**
     * From bytes.
     *
     * @param data the data
     * @return the endpoint object hash
     */
    public static EndpointObjectHash fromBytes(byte[] data) {
        if (data == null) {
            return null;
        }
        return new EndpointObjectHash(data);
    }

    /**
     * creates EndpointObjectHash using SHA1 algorithm over String representation of an object.
     *
     * @param data the data
     * @return the endpoint object hash
     */
    public static EndpointObjectHash fromSHA1(String data) {
        if (data == null) {
            return null;
        }
        return new EndpointObjectHash(SHA1HashUtils.hashToBytes(data));
    }

    /**
     * creates EndpointObjectHash using SHA1 algorithm over binary representation of an object.
     *
     * @param data the data
     * @return the endpoint object hash
     */
    public static EndpointObjectHash fromSHA1(byte[] data) {
        if (data == null) {
            return null;
        }
        return new EndpointObjectHash(SHA1HashUtils.hashToBytes(data));
    }

    /**
     * Gets the data of a hash.
     *
     * @return the data
     */
    public byte[] getData() {
        return Arrays.copyOf(data, data.length);
    }
    
    /**
     * Gets the byte buffer with a hash.
     *
     * @return the data
     */
    public ByteBuffer getDataBuf() {
        return ByteBuffer.wrap(getData());
    }

    /**
     * Checks if objects are binary equal.
     *
     * @param data the data
     * @return true, if successful
     */
    public boolean binaryEquals(byte[] data) {
        return Arrays.equals(this.data, data);
    }

    /* (non-Javadoc)
     * @see java.lang.Object#toString()
     */
    @Override
    public String toString() {
        return Arrays.toString(data);
    }

    /* (non-Javadoc)
     * @see java.lang.Object#hashCode()
     */
    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + Arrays.hashCode(data);
        return result;
    }

    /* (non-Javadoc)
     * @see java.lang.Object#equals(java.lang.Object)
     */
    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        EndpointObjectHash other = (EndpointObjectHash) obj;
        if (!Arrays.equals(data, other.data)) {
            return false;
        }
        return true;
    }
}
